File: src/aunmag/shooter/client/graphics/Hud.java
package aunmag.shooter.client.graphics;

import aunmag.nightingale.Input;
import aunmag.shooter.world.World;
import aunmag.nightingale.Application;
import aunmag.nightingale.data.DataEngine;
import aunmag.nightingale.font.Text;
import aunmag.nightingale.font.Font;
import aunmag.nightingale.utilities.UtilsMath;
import org.lwjgl.glfw.GLFW;

public class Hud {

    private static boolean isVisible = false;
    private static Text text = createText(DataEngine.titleFull);

    private static Text createText(String message) {
        return new Text(
                10,
                10,
                Application.getWindow().getWidth(),
                message,
                1f,
                Font.fontDefault,
                false
        );
    }

    public static void render() {
        if (Input.isKeyPressed(GLFW.GLFW_KEY_F1)) {
            isVisible = !isVisible;
        }

        if (!isVisible) {
            return;
        }

        float timeSpentUpdate = 0;
        float timeSpentRender = 0;
        float timeSpentTotal = timeSpentUpdate + timeSpentRender;
        float round = 100f;
        timeSpentUpdate = UtilsMath.calculateRoundValue(timeSpentUpdate, round);
        timeSpentRender = UtilsMath.calculateRoundValue(timeSpentRender, round);
        timeSpentTotal = UtilsMath.calculateRoundValue(timeSpentTotal, round);

        String message = "";
        message += String.format("Spent time on updating: %s ms\n", timeSpentUpdate);
        message += String.format("Spent time on rendering: %s ms\n", timeSpentRender);
        message += String.format("Spent time total: %s ms \n", timeSpentTotal);
        message += String.format("\nAIs: %s", World.ais.size());
        message += String.format("\nActors: %s", World.actors.size());
        message += String.format("\nBullets: %s", World.projectiles.size());
        message += String.format("\nTerrains: %s", World.terrains.size());
        message += String.format("\nDecorations: %s", World.decorations.size());

        if (!text.message.equals(message)) {
            text.delete();
            text = createText(message);
        }

        Font.fontDefault.renderPrepare();
        text.render();
    }

}

File: src/main/java/aunmag/shooter/ux/Hud.kt
package aunmag.shooter.ux

import aunmag.nightingale.font.FontStyleDefault
import aunmag.nightingale.font.Text
import aunmag.nightingale.input.Input
import aunmag.nightingale.utilities.UtilsMath
import aunmag.shooter.client.App
import aunmag.shooter.data.player
import aunmag.shooter.gui.Parameter
import org.lwjgl.glfw.GLFW

class Hud {

    private val health = Parameter("Health", 0.0f, 30, 32)
    private val stamina = Parameter("Stamina", 0.0f, 30, 33)
    private val ammo = Parameter("Ammo", 0.0f, 30, 34)
    private val debug = Text(10f, 10f, "", FontStyleDefault.simple)

    fun update() {
        Parameter.update()
        health.value = player?.health ?: 0f
        stamina.value = player?.stamina?.current ?: 0f
        ammo.value = player?.weapon?.magazine?.calculateVolumeRatio() ?: 0f
        ammo.isPulsing = player?.weapon?.magazine?.isReloading ?: false
    }

    fun render() {
        health.render()
        stamina.render()
        ammo.render()

        if (App.main.isDebug) {
            renderDebug()
        }
    }

    private fun renderDebug() {
        val game = App.main.game ?: return
        val world = game.world

        var timeSpentUpdate = 0f // TODO: Invoke data
        var timeSpentRender = 0f // TODO: Invoke data
        var timeSpentTotal = timeSpentUpdate + timeSpentRender
        val round = 100f
        timeSpentUpdate = UtilsMath.calculateRoundValue(timeSpentUpdate, round)
        timeSpentRender = UtilsMath.calculateRoundValue(timeSpentRender, round)
        timeSpentTotal = UtilsMath.calculateRoundValue(timeSpentTotal, round)

        var message = ""
        message += String.format("Spent time on updating: %s ms\n", timeSpentUpdate)
        message += String.format("Spent time on rendering: %s ms\n", timeSpentRender)
        message += String.format("Spent time total: %s ms \n", timeSpentTotal)
        message += String.format("\nAIs: %s", world.ais.size)
        message += String.format("\nActors: %s", world.actors.size)
        message += String.format("\nBullets: %s", world.projectiles.size)

        debug.load(message)
        debug.orderRendering()
    }

    fun remove() {
        health.remove()
        stamina.remove()
        ammo.remove()
        debug.remove()
    }

}